### PR TITLE
Add tracing to fs ops

### PR DESCRIPTION
**Pull request by kislaykishore**
### Description
* Add tracing to fs ops
* Increase signal to noise ratio by refactoring code.
* Annotating spans with error info will be handled in a subsequent PR.

image

### Link to the issue in case of a bug fix.
NA

### Testing details
1. Manual - NA
2. Unit tests - NA
3. Integration tests - NA

### CODE REVIEW

File: internal/fs/wrappers/monitoring.go [hunk 1]
```diff
@@ -29,12 +29,18 @@ import (
 	"go.opencensus.io/stats"
 	"go.opencensus.io/stats/view"
 	"go.opencensus.io/tag"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/codes"
 )
 
+const name = "github.com/googlecloudplatform/gcsfuse/fs"
```
Comment (charith87) [inline internal/fs/wrappers/monitoring.go:36]: We can change this to google's host. It shouldnt be github domain.

**Review by charith87 — COMMENTED** [review]
Pls check if it causes a memory leak or higher cpu as it is initializing always.

**Review by charith87 — APPROVED** [review]
Overall Approving to unblock. 
go/lgtm-with-comments

**Review by charith87 — DISMISSED** [review]

**Review by charith87 — APPROVED** [review]

File: internal/fs/wrappers/monitoring_test.go [hunk 2]
```diff
@@ -117,3 +124,148 @@ func TestFsErrStrAndCategory(t *testing.T) {
...
+func (d dummyFS) Destroy() {}
+
+func TestSpan(t *testing.T) {
+	ex := newInMemoryExporter(t)
+	m := monitoring{
+		wrapped: dummyFS{},
+	}
+
+	err := m.StatFS(context.Background(), nil)
+	require.NoError(t, err)
```
Comment (gargnitingoogle) [inline internal/fs/wrappers/monitoring_test.go:265]: how is this different from assert.NoError ? If no difference, let's use assert.NoError for consistency.

File: internal/fs/wrappers/monitoring_test.go [hunk 3]
```diff
@@ -117,3 +124,148 @@ func TestFsErrStrAndCategory(t *testing.T) {
...
+	return nil
+}
+
+func (d dummyFS) Fallocate(_ context.Context, _ *fuseops.FallocateOp) error {
+	return nil
+}
+
+func (d dummyFS) Destroy() {}
+
+func TestSpan(t *testing.T) {
```
Comment (gargnitingoogle) [inline internal/fs/wrappers/monitoring_test.go:258]: The name is a bit unclear to me. This sounds like it's testing the span functionality, which is not a gcsfuse feature.. How about TestInvocationSpan ?

File: internal/fs/wrappers/monitoring_test.go [hunk 4]
```diff
@@ -117,3 +124,148 @@ func TestFsErrStrAndCategory(t *testing.T) {
...
+
+func (d dummyFS) SetXattr(_ context.Context, _ *fuseops.SetXattrOp) error {
+	return nil
+}
+
+func (d dummyFS) Fallocate(_ context.Context, _ *fuseops.FallocateOp) error {
+	return nil
+}
+
+func (d dummyFS) Destroy() {}
```
Comment (gargnitingoogle) [inline internal/fs/wrappers/monitoring_test.go:256]: Given that Destroy is the only function of FS which is different from the rest in signature, and destroy is also interesting for deinitialization point of view, I suggest we add a test with it, or update the existing test with a call to `m.Destroy()` to see that it get recorded by monitoring.

**Comment by gargnitingoogle** [comment]
My bad. I guess this PR was open in  a tab in my browser for a few hours before I got to it, and didn't realize that it had been approved and merged already.

**Pull request merged**